Update XCM-Transactor and XCM-Fees page (#389)
PureStake/moonbeam-docs: builders/xcm/fees.md

## Relay Chain XCM Fee Calculation  {: #rel-chain-xcm-fee-calc }

Substrate has introduced a weight system that determines how heavy or, in other words, how expensive from a computational cost perspective an extrinsic is. One unit of weight is defined as one picosecond of execution time. When it comes to paying fees, users will pay a transaction fee based on the weight of the call that is being made, in addition to factors such as network congestion. 

The following sections will break down how to calculate XCM fees for Polkadot and Kusama. It's important to note that Kusama, in particular, uses benchmarked data to determine the total weight costs for XCM instructions and that some XCM instructions might include database reads/writes, which add weight to the call.

There are two databases available in Polkadot and Kusama, RocksDB (which is the default) and ParityDB, both of which have their own associated weight costs for each network. 

### Polkadot {: #polkadot }

> As previously mentioned, Polkadot currently uses a [fixed amount of weight](https://github.com/paritytech/polkadot/blob/e76cd144f9dad8c1304fd1476f92495bbb9ad22e/runtime/polkadot/src/xcm_config.rs#L95){target=_blank} for all XCM instructions, which is `{{ networks.polkadot.xcm_instruction.weight.transfer }}` weight units.

Reviewer: this shouldn't be changed, there is no corresponding variable for `weight.transfer`

| RocksDB (default) | {{ networks.polkadot.rocks_db.read_weight }}  | {{ networks.polkadot.rocks_db.write_weight }}  |
|     ParityDB      | {{ networks.polkadot.parity_db.read_weight }} | {{ networks.polkadot.parity_db.write_weight }} |

With the instruction weight cost established, you can calculate the cost of each instruction in DOT. 

In Polkadot, the [`ExtrinsicBaseWeight`](https://github.com/paritytech/polkadot/blob/master/runtime/polkadot/constants/src/weights/extrinsic_weights.rs#L56){target=_blank} is set to `{{ networks.polkadot.extrinsic_base_weight.display }}` which is [mapped to 1/10th](https://github.com/paritytech/polkadot/blob/master/runtime/polkadot/constants/src/lib.rs#L87){targer=blank} of a cent. Where 1 cent is `10^10 / 100`. 

Therefore, to calculate the cost of executing an XCM instruction, you can use the following formula:

```
XCM-DOT-COST = XCMInstrWeight * DOTWeightToFeeCoefficient
```

Where `DOTWeightToFeeCoefficient` is a constant (map to 1 cent), and can be calculated as:

```

> KSMWeightToFeeCoefficient = ( 10^10 / ( 10 * 100 )) * ( 1 / DOTExtrinsicBaseWeight )

Reviewer: ```suggestion
DOTWeightToFeeCoefficient = ( 10^10 / ( 10 * 100 )) * ( 1 / DOTExtrinsicBaseWeight )
```

```

Using the actual values:

```

> DOTWeightToFeeCoefficient = (10^10 / ( 10 * 100 * {{ networks.polkadot.extrinsic_base_weight.numbers_only }})

Reviewer: ```suggestion
DOTWeightToFeeCoefficient = ( 10^10 / ( 10 * 100 * {{ networks.polkadot.extrinsic_base_weight.numbers_only }} )
```

```

As a result, `DOTWeightToFeeCoefficient` is equal to `{{ networks.polkadot.xcm_instruction.planck_dot_weight }} Planck-DOT`. Now, you can begin to calculate the final fee in DOT, using `DOTWeightToFeeCoefficient` as a constant and `TotalWeight` as the variable:

```

> XCM-Planck-DOT-COST = TotalWeight * DOTWeightToFeeCoefficient

Reviewer: shouldn't `Planck` just be capitalized too? Either that or `COST` should be lower case?

Author: True, I've set `COST` to `Cost`

With the instruction weight cost established, you can calculate the cost of the instruction in KSM. 

In Kusama, the [`ExtrinsicBaseWeight`](https://github.com/paritytech/polkadot/blob/master/runtime/kusama/constants/src/weights/extrinsic_weights.rs#L56){target=_blank} is set to `{{ networks.kusama.extrinsic_base_weight.display }}` which is [mapped to 1/10th](https://github.com/paritytech/kusama/blob/master/runtime/polkadot/constants/src/lib.rs#L85){targer=blank} of a cent. Where 1 cent is `10^12 / 30,000`.

Therefore, to calculate the cost of executing an XCM instruction, you can use the following formula:

```
XCM-KSM-COST = XCMInstrWeight * KSMWeightToFeeCoefficient
```

Where `KSMWeightToFeeCoefficient` is a constant (map to 1 cent), and can be calculated as:

```

> KSMWeightToFeeCoefficient = ( 10^12 / ( 10 * 30000 )) * ( 1 / KSMExtrinsicBaseWeight )

Reviewer: I would do the same thing as you did in the DOT section here with the "Using the actual values:" and then showing the actual values

Author: nice catch!

| DepositAsset  | {{ networks.kusama.xcm_instruction.deposit_asset.total_weight }} | {{ networks.kusama.xcm_instruction.deposit_asset.ksm_cost }} KSM |
|   **TOTAL**   |      **{{ networks.kusama.xcm_message.transfer.weight }}**       |     **{{ networks.kusama.xcm_message.transfer.cost }} KSM**      |

## Moonbeam-based Networks XCM Fee Calculation  {: #moonbeam-xcm-fee-calc }

Substrate has introduced a weight system that determines how heavy or, in other words, how expensive an extrinsic is from a computational cost perspective. One unit of weight is defined as one picosecond of execution time. When it comes to paying fees, users will pay a transaction fee based on the weight of the call that is being made, and each parachain can decide how to convert from weight to fee, for example, accounting for additional costs for transaction size, and storage costs.

> The following sections will break down how to calculate XCM fees for Moonbeam-based networks. There are two main scenarios:

Reviewer: I think this makes more sense after the table 🤷‍♀️

Author: Agree

1. Fees paid in the reserve token (native tokens like GLMR, MOVR, or DEV)

> 2. Fees paid in external assets (XC-20s)

Reviewer: Not sure what you think, but I feel like we typically only use numbers for steps and bullet points for other lists... either way not really a big deal

Author: Yeah sure, you are more aware of the rules than me 😃

| [{{ networks.moonbeam.xcm.instructions.weight_units.display }}](https://github.com/PureStake/moonbeam/blob/f19ba9de013a1c789425d3b71e8a92d54f2191af/runtime/moonbeam/src/xcm_config.rs#L201){target=_blank} | [{{ networks.moonriver.xcm.instructions.weight_units.display }}](https://github.com/PureStake/moonbeam/blob/f19ba9de013a1c789425d3b71e8a92d54f2191af/runtime/moonriver/src/xcm_config.rs#L208){target=_blank} | [{{ networks.moonbase.xcm.instructions.weight_units.display }}](https://github.com/PureStake/moonbeam/blob/f19ba9de013a1c789425d3b71e8a92d54f2191af/runtime/moonbase/src/xcm_config.rs#L219){target=_blank} |

### Fee Calculation for Reserve Assets {: #moonbeam-reserve-assets }

> For each XCM instruction, the weight units are converted to balance units as part of the fee calculation. The amount of weight and Wei per weight for each of the Moonbeam-based networks is as follows:

Reviewer: ```suggestion
For each XCM instruction, the weight units are converted to balance units as part of the fee calculation. The amount of Wei per weight for each of the Moonbeam-based networks is as follows:
```

Which means that four XCM instructions cost `{{ networks.moonbeam.xcm.message.transfer.exec_time }}` seconds of block execution time.

To calculate the total cost in xcDOT, you'll also need the number of decimals the asset in question uses, which for xcDOT is ten decimals. You can determine the number of decimals for any asset by [querying the asset metadata](/builders/xcm/xc20/xc20/#x-chain-assets-metadata){target=_blank}.

The block execution formula can then be used to determine how much Alice's transfer of DOT to Alith's account on Moonbeam costs. The formula for finding the total cost is as follows:

```
XCM-Cost = (UnitsPerSecond / DecimalConversion) * ExecutionTime
```

Then the calculation for the transfer is:

```
XCM-Cost = ({{ networks.moonbeam.xcm.units_per_second.xcdot.transfer_numbers_only }} / 10^10) * {{ networks.moonbeam.xcm.message.transfer.exec_time }}
```

The total cost to transfer Alice's DOT to Alith's account for xcDOT is `{{ networks.moonbeam.xcm.message.transfer.xcdot_cost }} xcDOT`.

> ## XCM-Transactor Fees

Reviewer: ```suggestion
## XCM-Transactor Fees {: #xcm-transactor-fees }
```

> The [XCM-transactor pallet](/builders/xcm/xcm-transactor/) builds an XCM message to remotely transact in other chains of the ecosystem. 

Reviewer: ```suggestion
The [XCM-transactor pallet](/builders/xcm/xcm-transactor/){target=_blank} builds an XCM message to remotely transact in other chains of the ecosystem. 
```

There are two different ways developers can remotely transact through the pallet:

1. [`transactThroughDerivative`](/builders/xcm/xcm-transactor/#xcmtransactor-transact-through-derivative){target=_blank}
2. [`transactThroughSigned`](/builders/xcm/xcm-transactor/#xcmtransactor-transact-through-signed){target=_blank}, in which the dispatcher account in the destination chain is a multilocation-derived account and must have enough funds to cover XCM execution fees plus whatever call is being remotely executed

Generally speaking, the XCM instructions normally involved for remote execution are:

>  - Instructions to handle tokens in the origin chain. This can be either moving tokens to a sovereign account or burning the corresponding [XC-20](/builders/xcm/xc20/overview/){target=_blank} so that it can be used in the target chain. These instructions get executed in the origin chain

Reviewer: is this meant to be here? seems like it was just supposed to be a list of the actual instructions

Author: Yes, what I meant here is that the first instruction is always something to handle tokens in the origin chain. They differ depending if it is a Reserve asset or not. Any suggestions?

Reviewer: Maybe start with just that... "The first instruction handles tokens in the origin chain. This can be..."

Author: Thanks, I like that

 - [`WithdrawAsset`](https://github.com/paritytech/xcm-format#withdrawasset){target=_blank} - gets executed in the target chain. Removes assets and places them into the holding register
 - [`BuyExecution`](https://github.com/paritytech/xcm-format#buyexecution){target=_blank} - gets executed in the target chain. Takes the assets from holding to pay for execution fees. The fees to pay are determined by the target chain
 - [`Transact`](https://github.com/paritytech/xcm-format#transact){target=_blank} - gets executed in the target chain. Dispatches the encoded call data from a given origin

Therefore, XCM execution in the target chain consists of 3 to 4 XCM instructions, depending on the extrinsic being used. This section covers how XCM fees are estimated for each of the scenarios described above, as it is handled very differently.

> ### Transact Through Derivative Fees

Reviewer: ```suggestion
### Transact Through Derivative Fees {: #transact-through-derivative-fees }
```

Consider the following scenario: Alice wants to remotely transact in Polkadot from Moonbeam using the transact through sovereign extrinsic (she already has an index registered to her account). To estimate how many XC-20 tokens will be burned from Alice's account, you need to check the transact information specific to the relay chain. To do so, head to the chain state page of [Polkadot.js Apps](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fwss.api.moonbeam.network#/chainstate){target=_blank} and set the following options:

1. Choose the **xcmTransactor** pallet
2. Choose the **transactInfoWithWeightLimit** method
3. Set the multilocation for the destination chain from which you want to query the transact information. For this example, you can set `parents` to `1` and `interior` to `here`

> 4. Click on `+`

Reviewer: ```suggestion
4. Click on **+**
```

PICTURE_HERE

From the response, you can see that the `transactExtraWeight` is `{{ networks.polkadot.xcm_message.transact.weight }}`. This is the weight needed to execute the 3 XCM instructions for this remote call in that specific destination chain. Next, you need to find the `UnitsPerSecond` for that particular chain. In the same [Polkadot.js Apps page](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fwss.api.moonbeam.network#/chainstate){target=_blank}, set the following options:

1. Choose the **xcmTransactor** pallet
2. Choose the **destinationAssetFeePerSecond** method
3. Set the multilocation for the destination chain from which you want to query the transact information. For this example, you can set `parents` to `1` and `interior` to `here`

> 4. Click on `+`

Reviewer: ```suggestion
4. Click on **+**
```

PICTURE_HERE

> Note that this `UnitsPerSecond` is related to the cost estimated in the [Relay Chain XCM Fee Calculation](/builders/xcm/fees/#polkadot){target=_blank} section, or to the one shown in the [Units per weight](/builders/xcm/fees/#moonbeam-reserve-assets){target=_blank} section if the target is another parachain. As before, calculating the associated XCM execution fee is as simple as multiplying the `transactExtraWeight` times the `UnitsPerSecond`:

Reviewer: ```suggestion
Note that this `UnitsPerSecond` is related to the cost estimated in the [Relay Chain XCM Fee Calculation](#polkadot) section, or to the [Wei per weight](#moonbeam-reserve-assets) if the target is another parachain. As before, calculating the associated XCM execution fee is as simple as multiplying the `transactExtraWeight` times the `UnitsPerSecond`:
```

Consider the following scenario: Alice wants to remotely transact in another chain (Parachain ID 888, in the Moonbase Alpha relay chain ecosystem) from Moonbase Alpha using the transact through signed extrinsic. To estimate the amount of tokens Alice's multilocation-derivative account will need to have to execute the remote call, you need to check the transact information specific to the destination chain. To do so, head to the chain state page of [Polkadot.js Apps](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fwss.api.moonbeam.network#/chainstate){target=_blank} and set the following options:

1. Choose the **xcmTransactor** pallet
2. Choose the **transactInfoWithWeightLimit** method
3. Set the multilocation for the destination chain from which you want to query the transact information. For this example, you can set `parents` to `1` and `interior` to `X1: {"Parachain": 888}`

> 4. Click on `+`

Reviewer: ```suggestion
4. Click on **+**
```

PICTURE_HERE

From the response, you can see that the `transactExtraWeightSigned` is `{{ networks.moonbase_beta.xcm_message.transact.weight }}`. This is the weight needed to execute the 4 XCM instructions for this remote call in that specific destination chain. Next, you need to find how much the destination chain charges per weight of XCM execution. Normally, you would look into the `UnitsPerSecond` for that particular chain. But in this scenario, no XC-20 tokens are burned. Therefore, `UnitsPerSecond` can be use for reference, but do not ensure that the amount of tokens estimated are correct. To get the `UnitsPerSecond` as a reference value, in the same [Polkadot.js Apps page](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fwss.api.moonbeam.network#/chainstate){target=_blank}, set the following options:

1. Choose the **xcmTransactor** pallet
2. Choose the **destinationAssetFeePerSecond** method
3. Set the multilocation for the destination chain from which you want to query the transact information. For this example, you can set `parents` to `1` and `interior` to `X2: [{"Parachain": 888},{"PalletInstance":3}]`

> 4. Click on `+`

Reviewer: ```suggestion
4. Click on **+**
```
